File: urirang/app/src/main/java/com/makeus/urirang/android/src/ApplicationClass.java
package com.makeus.urirang.android.src;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.iid.FirebaseInstanceId;
import com.google.firebase.iid.InstanceIdResult;
import com.kakao.auth.ApprovalType;
import com.kakao.auth.AuthType;
import com.kakao.auth.IApplicationConfig;
import com.kakao.auth.ISessionConfig;
import com.kakao.auth.KakaoAdapter;
import com.kakao.auth.KakaoSDK;
import com.makeus.urirang.android.config.XAccessTokenInterceptor;

import java.text.DecimalFormat;
import java.text.SimpleDateFormat;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ApplicationClass extends Application {
    public static MediaType MEDIA_TYPE_JSON = MediaType.parse("application/json; charset=uft-8");
    public static MediaType MEDIA_TYPE_JPEG = MediaType.parse("image/jpeg");

    //서버 주소
    public static String BASE_URL = "http://ec2-13-124-41-220.ap-northeast-2.compute.amazonaws.com/";

    //공용 shared preference
    public static SharedPreferences sSharedPreferences = null;

    // 가격 포멧
    public static DecimalFormat myFormatter = new DecimalFormat("###,###");

    // SharedPreferences 키 값
    public static String TAG = "BreezeWind";

    // JWT Token 키 값
    public static final String X_ACCESS_TOKEN = "X-ACCESS-TOKEN";
    // FCM Token 키 값
    public static final String FCM_TOKEN = "URIRANG_FCM_TOKEN";

    //날짜 형식
    public static SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd", Locale.KOREA);
    public static SimpleDateFormat HOME_MONTH = new SimpleDateFormat("MM", Locale.KOREA);
    public static SimpleDateFormat HOME_DAY = new SimpleDateFormat("dd", Locale.KOREA);

    // Retrofit 인스턴스
    public static Retrofit retrofit;
    public static Retrofit retrofitForKakao;

    private static ApplicationClass instance;

    public static ApplicationClass getApplicationClassContext() {
        if (instance == null) {
            throw new IllegalStateException("This Application does not inherit com.kakao.GlobalApplication");
        }
        return instance;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;

        if (sSharedPreferences == null) {
            sSharedPreferences = getApplicationContext().getSharedPreferences(TAG, Context.MODE_PRIVATE);
        }

        KakaoSDK.init(new KakaoSDKAdapter());

    }

    @Override
    public void onTerminate() {
        super.onTerminate();
        instance = null;
    }

    public static void getFcmToken() {
        FirebaseInstanceId.getInstance().getInstanceId().addOnCompleteListener(new OnCompleteListener<InstanceIdResult>() {
            @Override
            public void onComplete(@NonNull Task<InstanceIdResult> task) {
                if (!task.isSuccessful()) {
                    Log.d("로그", "getInstanceId failed", task.getException());
                    return;
                }

                String token = task.getResult().getToken();
//                Log.d(TAG, "fcm: " + token);
                SharedPreferences.Editor editor = sSharedPreferences.edit();
                editor.putString(FCM_TOKEN, token);
                editor.apply();
            }
        });
    }

    public static Retrofit getRetrofit() {
        if (retrofit == null) {
            OkHttpClient client = new OkHttpClient.Builder()
                    .readTimeout(5000, TimeUnit.MILLISECONDS)
                    .connectTimeout(5000, TimeUnit.MILLISECONDS)
                    .addNetworkInterceptor(new XAccessTokenInterceptor()) // JWT 자동 헤더 전송
                    .build();

            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }

        return retrofit;
    }

    public static Retrofit getRetrofitForImageUpload() {
        if (retrofit == null) {
            OkHttpClient client = new OkHttpClient.Builder()
                    .readTimeout(30000, TimeUnit.MILLISECONDS)
                    .connectTimeout(30000, TimeUnit.MILLISECONDS)
                    .addNetworkInterceptor(new XAccessTokenInterceptor()) // JWT 자동 헤더 전송
                    .build();

            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }

        return retrofit;
    }

    public static Retrofit getRetrofitForKakao() {
        if (retrofitForKakao == null) {
            OkHttpClient client = new OkHttpClient.Builder()
                    .readTimeout(5000, TimeUnit.MILLISECONDS)
                    .connectTimeout(5000, TimeUnit.MILLISECONDS)
                    .build();

            retrofitForKakao = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }

        return retrofitForKakao;
    }

    public class KakaoSDKAdapter extends KakaoAdapter {

        @Override
        public IApplicationConfig getApplicationConfig() {
            return new IApplicationConfig() {
                @Override
                public Context getApplicationContext() {
                    return ApplicationClass.getApplicationClassContext();
                }
            };
        }

        @Override
        public ISessionConfig getSessionConfig() {
            return new ISessionConfig() {

                @Override
                public AuthType[] getAuthTypes() {
                    return new AuthType[]{AuthType.KAKAO_TALK};
                }

                @Override
                public boolean isUsingWebviewTimer() {
                    return false;
                }

                @Override
                public boolean isSecureMode() {
                    return false;
                }

                @Nullable
                @Override
                public ApprovalType getApprovalType() {
                    return ApprovalType.INDIVIDUAL;
                }

                @Override
                public boolean isSaveFormData() {
                    return true;
                }
            };
        }
    }


}

File: urirang/app/src/main/java/com/makeus/urirang/android/src/ApplicationClass.kt
package com.makeus.urirang.android.src

import android.app.Application
import android.content.SharedPreferences
import com.kakao.auth.*
import com.makeus.urirang.android.config.XAccessTokenInterceptor
import okhttp3.MediaType
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.text.DecimalFormat
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit

class ApplicationClass : Application() {
    override fun onCreate() {
        super.onCreate()
        instance = this
        if (sSharedPreferences == null) {
            sSharedPreferences = applicationContext.getSharedPreferences(TAG, MODE_PRIVATE)
        }

//        KakaoSDK.init(new KakaoSDKAdapter());
    }

    override fun onTerminate() {
        super.onTerminate()
        instance = null
    }

    inner class KakaoSDKAdapter : KakaoAdapter() {
        override fun getApplicationConfig(): IApplicationConfig {
            return IApplicationConfig { applicationClassContext!! }
        }

        override fun getSessionConfig(): ISessionConfig {
            return object : ISessionConfig {
                override fun getAuthTypes(): Array<AuthType> {
                    return arrayOf(AuthType.KAKAO_TALK)
                }

                override fun isUsingWebviewTimer(): Boolean {
                    return false
                }

                override fun isSecureMode(): Boolean {
                    return false
                }

                override fun getApprovalType(): ApprovalType? {
                    return ApprovalType.INDIVIDUAL
                }

                override fun isSaveFormData(): Boolean {
                    return true
                }
            }
        }
    }

    companion object {
        //서버 주소
        var BASE_URL = "http://49.50.174.243"

        //공용 shared preference
        @JvmField
        var sSharedPreferences: SharedPreferences? = null

        // 가격 포멧
        var myFormatter = DecimalFormat("###,###")

        // SharedPreferences 키 값
        var TAG = "revelyview"

        // JWT Token 키 값
        const val X_ACCESS_TOKEN = "X-ACCESS-TOKEN"

        // FCM Token 키 값
        const val FCM_TOKEN = "OMO_FCM_TOKEN"

        //날짜 형식
        var DATE_FORMAT = SimpleDateFormat("yyyy-MM-dd", Locale.KOREA)
        var HOME_MONTH = SimpleDateFormat("MM", Locale.KOREA)
        var HOME_DAY = SimpleDateFormat("dd", Locale.KOREA)

        // Retrofit 인스턴스
        var retrofit: Retrofit? = null
        var retrofitForKakao: Retrofit? = null
        private var instance: ApplicationClass? = null
        val applicationClassContext: ApplicationClass?
            get() {
                checkNotNull(instance) { "This Application does not inherit com.kakao.GlobalApplication" }
                return instance
            }

        fun getRetrofit(): Retrofit? {
            if (retrofit == null) {
                val client = OkHttpClient.Builder()
                        .readTimeout(5000, TimeUnit.MILLISECONDS)
                        .connectTimeout(5000, TimeUnit.MILLISECONDS)
                        .addNetworkInterceptor(XAccessTokenInterceptor()) // JWT 자동 헤더 전송
                        .build()
                retrofit = Retrofit.Builder()
                        .baseUrl(BASE_URL)
                        .client(client)
                        .addConverterFactory(GsonConverterFactory.create())
                        .build()
            }
            return retrofit
        }

        fun getRetrofitForKakao(): Retrofit? {
            if (retrofitForKakao == null) {
                val client = OkHttpClient.Builder()
                        .readTimeout(5000, TimeUnit.MILLISECONDS)
                        .connectTimeout(5000, TimeUnit.MILLISECONDS)
                        .build()
                retrofitForKakao = Retrofit.Builder()
                        .baseUrl(BASE_URL)
                        .client(client)
                        .addConverterFactory(GsonConverterFactory.create())
                        .build()
            }
            return retrofitForKakao
        }
    }
}